src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1229,50 +1484,29 @@ CHIP_ERROR CASESession::SendSigma3()
 
     ChipLogProgress(SecureChannel, "Sent Sigma3 msg");
 
-    err = mCommissioningHash.Finish(messageDigestSpan);
-    SuccessOrExit(err);
+    {
+        MutableByteSpan messageDigestSpan(mMessageDigest);
+        SuccessOrExit(err = mCommissioningHash.Finish(messageDigestSpan));
+    }
 
     mState = State::kSentSigma3;
 
 exit:
+    mSendSigma3Helper.reset();
 
-    if (err != CHIP_NO_ERROR)
+    // If data.keystore is set, processing occurred in the background, so if an error occurred,
+    // need to send status report (normally occurs in SendSigma3a), and discard exchange and
+    // abort pending establish (normally occurs in OnMessageReceived).
+    if (data.keystore != nullptr && err != CHIP_NO_ERROR)
     {
         SendStatusReport(mExchangeCtxt, kProtocolCodeInvalidParam);
```

Comment:
Followup (because this is a problem with the existing async Sigma3 handling code too): If this SendStatusReport call fails to send, we are leaking the exchange.  We need to fix that.

Reply:
Noted in issue #26280.